package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c

```diff
@@ -413,6 +415,70 @@ static void g997_line_inventory(int fd) {
 	m_array("serial", out.data.SerialNumber, DSL_G997_LI_MAXLEN_SERIAL);
 }
 
+static void g977_get_bit_allocation(int fd, DSL_AccessDir_t direction) {
+	IOCTL_DIR(DSL_G997_BitAllocationNsc_t, DSL_FIO_G997_BIT_ALLOCATION_NSC_GET, direction);
+	m_u32("groupsize", 1);  // create default value to obtain consistent JSON structure
+	m_u32("groups", out.data.bitAllocationNsc.nNumData);
+	m_array("data", &out.data.bitAllocationNsc.nNSCData, out.data.bitAllocationNsc.nNumData);
+}
+
+
+static void g977_get_snr(int fd, DSL_AccessDir_t direction) {
+	IOCTL_DIR_DELT(DSL_G997_DeltSnr_t, DSL_FIO_G997_DELT_SNR_GET, direction, DSL_DELT_DATA_SHOWTIME);
+	m_u32("groupsize", out.data.nGroupSize);
+	m_u32("groups", out.data.deltSnr.nNumData);
+
+	void *c = blobmsg_open_array(&b, "data");
+	
+	for (uint16_t i  = 0  ; i < out.data.deltSnr.nNumData ; i++) { //uint16_t i = 0; i < len; ++i
+		if (out.data.deltSnr.nNSCData[i] != 255 && out.data.deltSnr.nNSCData[i] != 255 != 0) {
+			m_double("", -32 + (double)out.data.deltSnr.nNSCData[i] / 2); // SNR -32 ... 95 dB
+		} else {
+			m_str("", "NaN");
```

Comment:
github sucks, from the other discussion marked as resolved:
That looks better, but I meant the actual NAN define/constant from math.h instead of a string. Does that work?

Reply:
On the NaN topic, that might not be trivial. Doing m_double("", NAN), results in `nan` being output in the json file. That would be OK but `nan` is considered by json as a string and should appear in between "" hence resulting in invalid json. If m_double("", NAN) would result in `"null"` as output that would at least result in valid json. Conclusion: for now not sure whether it adds any value to change.

Reply:
Alright, thanks for checking!
What I'm trying to avoid is to have every consumer having to special case whatever we special case here ;)
So if there's something for ubus/json we can use to represent these invalid values, we should use it.
So I guess `NaN` is out of the picture, in fact it's not part of the json spec and there's a [whole python issue](https://github.com/python/cpython/issues/84813) about it.

Stemming from that discussion is json's `null`. Can you please try that instead? It looks like it's represented by libubus' `BLOBMSG_TYPE_UNSPEC`, which we can use in `m_double`. Something like
```
static inline void m_double(const char *id, double value) {
	if (value = NAN)
		blobmsg_add_field(&b, BLOBMSG_TYPE_UNSPEC, id, NULL, 0);
	else
		blobmsg_add_double(&b, id, value);
}
```

And then just `m_double(id, NAN)` for those values. The json should be a valid `null` then.

Reply:
The snipped above doesn't work, but it works on my branch I mentioned below